reed-solomon-novelpoly/src/novel_poly_basis/mod.rs

```diff
@@ -231,6 +233,52 @@ impl ReedSolomon {
 
 		Ok(acc)
 	}
+
+	/// Reconstruct from the set of systematic chunks.
+	/// Systematic chunks are the first `k` chunks, which contain the initial data.
+	///
+	/// Provide a vector containing chunk data. If too few chunks are provided, recovery is not
+	/// possible.
+	/// The result may be padded with zeros. Truncate the output to the expected byte length.
+	pub fn reconstruct_from_systematic<S: Shard>(&self, chunks: Vec<S>) -> Result<Vec<u8>> {
+		let Some(first_shard) = chunks.first() else {
+			return Err(Error::NeedMoreShards { have: 0, min: self.k, all: self.n });
+		};
+		let shard_len = AsRef::<[[u8; 2]]>::as_ref(first_shard).len();
+
+		if shard_len == 0 {
+			return Err(Error::InconsistentShardLengths { first: 0, other: 0 });
+		}
+
+		if let Some(length) = chunks.iter().find_map(|c| {
+			let length = AsRef::<[[u8; 2]]>::as_ref(c).len();
+			if length != shard_len {
+				Some(length)
+			} else {
+				None
+			}
+		}) {
+			return Err(Error::InconsistentShardLengths { first: shard_len, other: length });
```

Comment:
I just realized this is being checked 3 times: here, in erasure-coding crate and in availability-recovery..

Reply:
you're right, I'll move all checks here. I only see it being checked here and in erasure-coding crate. Where is it being checked in availability-recovery?

Reply:
I'll remove checks from polkadot once this is merged and released. The only check that needs to remain in polkadot is for the even length of shards. This is needed because WrappedShard will panic if the length is not even and this function expects a WrappedShard.

Reply:
> Where is it being checked in availability-recovery?

I meant in your systematic recovery PR, but it will probably use this function instead, so nevermind.

> The only check that needs to remain in polkadot is for the even length of shards. This is needed because WrappedShard will panic if the length is not even and this function expects a WrappedShard.

I see, thanks for the clarification.

Reply:
> I meant in your systematic recovery PR, but it will probably use this function instead, so nevermind.

Ah, yes, I'll switch my PR to use this function